### PR TITLE
[AIR] `MultiHotEncoder` and list support for encoders

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR adds support for list columns for existing encoders (with `OrdinalEncoder` being able to both encode lists and to treat them as separate categories), as well as a new `MultiHotEncoder` preprocessor that allows for multi-label binarization of list columns.

As an aside, there doesn't seem to be a standard name for this in literature. sklearn uses `MultiLabelBinarizer`, but `MultiHotEncoder` is also seen, as are other names.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/preprocessors/encoder.py [hunk 1]
```diff
@@ -88,33 +119,121 @@ class OneHotEncoder(Preprocessor):
...
+    >>> )
+    >>> mhe.fit(ray.data.from_pandas(batch))
+    >>> transformed_batch = mhe.transform_batch(batch)
+    >>> expected_batch = pd.DataFrame(
+    >>>     {
+    >>>         "A": [[0, 0, 1], [0, 0, 0], [1, 1, 1], [2, 0, 0]],
+    >>>         "B": [[0, 0, 1], [1, 0, 0], [0, 1, 0], [1, 0, 0]],
+    >>>     }
+    >>> )
+    >>> assert transformed_batch.equals(expected_batch)
```
Comment (richardliaw) [inline python/ray/ml/preprocessors/encoder.py:183]: Thanks for adding this example!

File: python/ray/ml/preprocessors/encoder.py [hunk 2]
```diff
@@ -15,33 +17,58 @@ class OrdinalEncoder(Preprocessor):
...
 
+    All column values must be hashable scalars or lists of hashable values.
+
     Args:
         columns: The columns that will individually be encoded.
+        encode_lists: If True, each element of lists inside list
+            columns will be encoded. If False, each list will
+            be treated as a whole separate category. True
+            by default.
     """
```
Comment (xwjiang2010) [inline python/ray/ml/preprocessors/encoder.py:28]: Could we add an example snippet here? with different `encode_lists` options?

File: python/ray/ml/preprocessors/encoder.py [hunk 3]
```diff
@@ -15,33 +17,58 @@ class OrdinalEncoder(Preprocessor):
 
     Transforming values not included in the fitted dataset will be encoded as ``None``.
 
+    All column values must be hashable scalars or lists of hashable values.
```
Comment (xwjiang2010) [inline python/ray/ml/preprocessors/encoder.py:20]: Should we call out the assumption here? A column can be either all scalars or all lists. Cannot be a mix?

File: python/ray/ml/preprocessors/encoder.py [hunk 4]
```diff
@@ -88,33 +119,121 @@ class OneHotEncoder(Preprocessor):
...
+
+    Example:
+
+    >>> import ray.data
+    >>> from ray.ml.preprocessors import MultiHotEncoder
+    >>> import pandas as pd
+    >>> mhe = MultiHotEncoder(columns=["A", "B"])
+    >>> batch = pd.DataFrame(
+    ...     {
+    ...         "A": [["warm"], [], ["hot", "warm", "cold"], ["cold", "cold"]],
```
Comment (xwjiang2010) [inline python/ray/ml/preprocessors/encoder.py:171]: Love the example here!
Do you want to change ["warm"] to just "warm"?

**Review by xwjiang2010 — APPROVED** [review]
Thanks for adding it, Antoni!

File: python/ray/ml/preprocessors/encoder.py (hunk 3 above)
Comment (Yard1) [reply python/ray/ml/preprocessors/encoder.py:20]: No, it cannot. I am not sure if arrow even allows that.

File: python/ray/ml/preprocessors/encoder.py (hunk 4 above)
Comment (Yard1) [reply python/ray/ml/preprocessors/encoder.py:171]: Should be all lists

Comment (xwjiang2010) [reply python/ray/ml/preprocessors/encoder.py]: Then we probably should call out the assumption that no mixing of scaler and list is expected within one column right? The same goes for the OrdinalEncoder. It's easy for user to mix it up and no error is thrown.

Comment (Yard1) [reply python/ray/ml/preprocessors/encoder.py]: I updated the docstrings

**Pull request merged**